### PR TITLE
Refactor gzip downloader

### PR DESCRIPTION
**Pull request by HaozhouWang**

### CODE REVIEW

File: gpAux/extensions/gps3ext/src/uncompress_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// Read compressed data from underlying reader and uncompress to this->out buffer.
+// If no more data to consume, this->zstream.avail_out == S3_ZIP_CHUNKSIZE;
+void UncompressReader::uncompress() {
+    if (this->zstream.avail_in == 0) {
+        this->zstream.avail_out = S3_ZIP_CHUNKSIZE;
+        this->zstream.next_out = (Byte *)this->out;
+
+        // 1. reader S3_ZIP_CHUNKSIZE data from underlying reader and put into this->in buffer.
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/uncompress_reader.cpp]: Where is the "2."? :)

File: gpAux/extensions/gps3ext/src/uncompress_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+        this->zstream.next_in = (Byte *)this->in;
+        this->zstream.avail_in = hasRead;
+    } else {
+        // Still have more data in 'in' buffer to decode.
+        this->zstream.avail_out = S3_ZIP_CHUNKSIZE;
+        this->zstream.next_out = (Byte *)this->out;
+    }
+
+    S3DEBUG("Before decompress: avail_in = %d, avail_out = %d, total_in = %d, total_out = %d",
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/uncompress_reader.cpp]: Better to check if it's an int, same as other log printings in this file.

File: gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "uncompress_reader.cpp"
+#include "uncompress_reader.h"
+
+using std::vector;
+
+#define CHECK_ERR(err, msg)                              \
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp]: This function could be retired.

File: gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    {                                                    \
+        if (err != Z_OK) {                               \
+            fprintf(stderr, "%s error: %d\n", msg, err); \
+            exit(1);                                     \
+        }                                                \
+    }
+
+class BufferReader : public Reader {
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp]: Better to name it MockBufferReader?

File: gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+    char hello[S3_ZIP_CHUNKSIZE * 2 + 2];
+    memset((void *)hello, 'A', sizeof(hello));
+    hello[sizeof(hello) - 1] = '\0';
+
+    setBufReaderByRawData(hello, sizeof(hello));
+
+    char outputBuffer[16] = {0};
+
+    // read 1st 16 bytes
+    S3DEBUG("1st");
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp]: "// read 1st 16 bytes" and "// read 1st 16 bytes" are not necessary, also other places in this file.

File: gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    // TearDown() is invoked immediately after a test finishes.
+    virtual void TearDown() {
+        uncompressReader.close();
+    }
+
+    void setBufReaderByRawData(const char *input, int len) {
+        uLong compressedLen = sizeof(compressionBuff);
+
+        int err = compress(compressionBuff, &compressedLen, (const Bytef *)input, len);
+        CHECK_ERR(err, "failed to compress sample data");
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp]: Better switch to S3DEBUG().

File: gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+        bufReader.setData(compressionBuff, compressedLen);
+    }
+
+    void enableDebug() {
```
Comment (adam8157) [inline gpAux/extensions/gps3ext/test/uncompress_reader_test.cpp]: Better to restore the orig loglevel and logtype once that very case finished.

**Comment by HaozhouWang** [comment]
Thanks for comment, has fixed.

**Pull request merged**